analytic_engine/src/instance/open.rs

```diff
@@ -138,71 +139,265 @@ impl Instance {
     }
 
     /// Open the table.
-    pub async fn do_open_table(
+    pub async fn do_open_tables_of_shard(
         self: &Arc<Self>,
-        space: SpaceRef,
-        request: &OpenTableRequest,
-    ) -> Result<Option<TableDataRef>> {
-        if let Some(table_data) = space.find_table_by_id(request.table_id) {
-            return Ok(Some(table_data));
+        context: TablesOfShardContext,
+    ) -> Result<OpenTablesOfShardResult> {
+        let mut shard_opener = ShardOpener::init(
+            context,
+            self.space_store.manifest.clone(),
+            self.space_store.wal_manager.clone(),
+            self.replay_batch_size,
+            self.make_flusher(),
+            self.max_retry_flush_limit,
+        )?;
+
+        shard_opener.open().await
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct TablesOfShardContext {
+    /// Shard id
+    pub shard_id: ShardId,
+    /// Table infos
+    pub table_ctxs: Vec<TableContext>,
+}
+
+#[derive(Clone, Debug)]
+pub struct TableContext {
+    pub table_def: TableDef,
+    pub space: SpaceRef,
+}
+
+#[derive(Debug)]
+enum TableOpenState {
+    RecoverTableMeta(RecoverTableMetaContext),
+    RecoverTableData(RecoverTableDataContext),
+    Failed(crate::instance::engine::Error),
+    Success(Option<SpaceAndTable>),
+}
```

Comment:
How about putting it:
```suggestion
enum TableOpenStage {
    Ready(Option<SpaceAndTable>),
    TableMetaRecovered(RecoverTableMetaContext),
    TableDataRecovered(RecoverTableDataContext),
    Failed(crate::instance::engine::Error),
    Success(Option<SpaceAndTable>),
}
```

The reason why I'm in favor of `Stage` is that there is no complex transition between these states and `Stage` suits to the linear transition.